rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,346 @@
+                    event.add_batch_notifier(Arc::clone(&batch));
+                }
+                receiver
+            });
+
+            out.send_all(&mut futures::stream::iter(events).map(Ok))
+                ...
+                .and_then(|_| handle_batch_status(ack_id, receiver))
+                .await
+    }
+    ...
+    async fn handle_batch_status(ack_id, receiver) {
+        match receiver.await {
+            BatchStatus::Delivered => // update ackId -> true,
+            BatchStatus::Errored => // leave ackId -> false,
+            BatchStatus::Failed => // leave ackId -> false,
+        }
+    }
+    ```
+
+  * When the data in the request is successfully delivered, we mark
+    `ack_ids_ack_status[currently_available_ack_id] = 1`.
+* Querying ackId status
+  * We add a new `/services/collector/ack` route
+  * `ackId`’s from incoming requests are used to index into
+    `ack_ids_ack_status`. We return true/false depending on the value in the
+    bitvec.
+  * If we return true for an ackId, we mark `ack_ids_in_use[ack_id] = 0` and
+    `ack_ids_ack_status[ack_id] = 0`. The `ackId` can then be reused.
+
+## Proposal: `splunk_hec` Sinks Indexer Acknowledgement
+
+### User Experience
+
+The `splunk_hec` sinks will automatically integrate with indexer
+acknowledgements if the user has enabled it in their Splunk instance (if
+`ackId`’s are present in HEC responses). If indexer acknowledgement is disabled,
+`splunk_hec` sinks will continue to finalize events based on the HEC response
+status code.
+
+### Implementation
+
+* [Splunk recommendations for client integration with indexer acknowledgement](https://docs.splunk.com/Documentation/Splunk/8.2.3/Data/AboutHECIDXAck#Indexer_acknowledgment_client_behavior)
+
+We generate a single `guid` to use as a channel ID and include this in HEC
+requests. After submitting events to HEC, we will parse the HTTP response for an
+`ackId`. If no `ackId` is found, we rely on the current behavior of setting
+`EventStatus` based solely on the HTTP status code.
+
+If an `ackId` is found, we store it in a `Map<u64, bool>` shared with a
+background tokio task which, for all pending `ackId`’s, will query
+`/services/collector/ack`. This background task will query at an interval of 10
+seconds (as recommended by Splunk) and update the shared structure based on the
+results. Meanwhile, back in the HEC response handler, we’ll check the shared
+structure for `true` at 10 second intervals up to a maximum time limit (5
+minutes as recommended by Splunk).
+
+```{rust}
+fn call(&mut self, req: HecRequest) -> Self::Future {
+        let mut http_service = self.batch_service.clone();
+        Box::pin(async move {
+            ...
+            // handle response
+            let response = http_service.call(req).await?;
+            // if ack_id is found
+            self.ack_id_to_status_map.insert(ack_id, false);
+            let mut interval = time::interval(Duration::from_secs(10));
```

Comment:
This logic appears to set up one polling loop _per-request_, which is not the behavior I read into the above description, and could lead to pile-ups. Could you clarify?

Reply:
Sure thing. There will be a loop per-request that checks whether the `ackId` status has been updated. A separate task (not shown in the pseudo-code) will actually be querying `/services/collector/ack` and updating `ackId` status(es) based on the response (for all the `ackId`'s we're currently tracking). The loop per-request will quit checking after a fixed amount of retries and assume the data has been dropped. Does that make sense?

The "pile-up" issue is definitely a concern; I'm not sure what the best way to fix that would be other than more strictly limiting retries.

Reply:
Ok, I misread this. I think what you want here is to:
1. Create a oneshot notifier for each request.
2. Add the `ackId` to a shared map tied to the send end of that notifier and a timeout.
3. Simply `await` on the receive end here instead of looping.
4. Set up a background polling task to check all the `ackId`s and time out expired entries, and send notifications based on that status.

Reply:
Thanks! Updated in 14e7e4d to the suggested design